### ISSUE TITLE
stop attributes for taxi device should be configurable

### ISSUE THREAD

**Issue by behrisch** [eclipse/sumo#6714]
especially street stop vs parking but maybe also position etc.

**Comment by namdre** [comment]
Setting a default could be done via a **--device.taxi.** option. However, it might be necessary to configure this per-location (some edges allow parking while others require on-street stopping).

**Comment by ArmelliniMG** [comment]
When working with traci, it would be important to define the stops with "expected and permitted" attributes to ensure that the vehicle will wait to the assigned person, and that only that person may enter the vehicle (similar as #7869).

**Comment by namdre** [comment]
> When working with traci, it would be important to define the stops with "expected and permitted" attributes to ensure that the vehicle will wait to the assigned person, and that only that person may enter the vehicle (similar as #7869).

re: permitted
In the current implementation, only customers of the current dispatch sequence may enter the taxi (this is like using permitted but for all customers of the sequence). Attribute 'permitted' on individual stops would only have an effect if the dispatch is looped and the customer should not be picked up on the first loop. Do we have a use case for this? 

re: expected
Currently, reservations are only visible once the person is waiting to be picked up. This means, attribute 'expected' should never take effect (though it may be needed in the future for advance reservations and late customers).

**Comment by ArmelliniMG** [comment]
re: permitted
Yes, it would be necessary regarding the capacity limits of the vehicle. Also, in most scenarios, waiting at stops/home is prefereable as riding. So it is relevant that the passenger gets on the taxi at the scheduled pickup time.

re: expected
You are right, it is not needed for the online version :)

**Comment by namdre** [comment]
@ArmelliniMG I've moved your request to #8295 since this is not about optional configuration but rather about correctness.

**Comment by namdre** [comment]
configured dropOffDuration doesn't work yet because a new dispatch simply aborts the current stop regardless of remaining duration.

**Comment by rummel123** [comment]
> 
> 
> Setting a default could be done via a **--device.taxi.** option. However, it might be necessary to configure this per-location (some edges allow parking while others require on-street stopping).

As a first step, if using busStops for the taxi stops, the value of an additional parking attribute of the busStop could be used to decide whether the taxi stops on the edge or next to it.
A general parking attribute for edges does not seem to fit the general requirements to a taxi stop. Then it became more meaningful to automatical use available parking areas next to the edge.

**Comment by namdre** [comment]
> As a first step, if using busStops for the taxi stops...

At the moment, taxis never stop at a busStop. They always stop via lane,pos positioning. This holds true even if a person is waiting at a busStop.

**Comment by denharma** [comment]
I want to highlight this issue again due to our current work within the ViVre-Project. We are really looking forward to a possibility, to individualize the stop attribute for the taxi-device. It would be great to be able to decide for each stop position (or even for each edge or lane) whether the taxi should stop on the edge or next to the edge.

**Comment by namdre** [comment]
suggestion:
- new option **--device.taxi.stop-onroad** and  `<param key="device.taxi.stop-onroad" .../>` to change the default from parking="true" to parking="false"
- if stop-onroad=true and there is a free parkingArea in range of the stop, that parkingArea is used instead

by using the new option and adding parkingAreas the user can control exactly where taxis stop on the road and where they leave the road.

@rummel123 @denharma  would that work for you?

**Comment by denharma** [comment]
Yes, sounds good, I think that would definitely help us. Thanks!

**Comment by namdre** [comment]
Part of the behavior is already working in 1.11 (see updated documentation at https://sumo.dlr.de/docs/Simulation/Taxi.html#customer_stops)

**Comment by namdre** [comment]
opportunistic parking cannot be modeled by the taxi device alone because the availability of a free parkingArea is only known when arriving at the stop location (whereas the taxi device defines stop attributes at dispatch time). It must be solved via #9691

**Comment by namdre** [comment]
@denharma here is how you can model location-based off-road parking already:

- set `<param key="device.taxi.parking" value="false"/>`
- in places where stopping taxis should not block the road, add an extra (dead-end) lane with `allow="taxi"`
